HW2/P3/mandelbrot.pyx

```diff
@@ -42,7 +45,89 @@ cpdef mandelbrot(np.complex64_t [:, :] in_coords,
                     z = z * z + c
                 out_counts[i, j] = iter
 
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cpdef mandelbrot2(np.complex64_t [:, :] in_coords,
+                 np.uint32_t [:, :] out_counts,
+                 int max_iterations=511):
+    cdef:
+       int i, j, iter, mask_signs, ind
+       AVX.float8 cnew_r, cnew_i, znew_r, znew_i, mags, iters, ones, fours, mask, twos, znew_i_tmp, znew_r_tmp
+       np.complex64_t c, z
+       float out_vals[8]
+       float [:] out_view = out_vals
+
+       # To declare AVX.float8 variables, use:
+       # cdef:
+       #     AVX.float8 v1, v2, v3
+       #
+       # And then, for example, to multiply them
+       #     v3 = AVX.mul(v1, v2)
+       #
+       # You may find the numpy.real() and numpy.imag() fuctions helpful.
+
+    assert in_coords.shape[1] % 8 == 0, "Input array must have 8N columns"
+    assert in_coords.shape[0] == out_counts.shape[0], "Input and output arrays must be the same size"
+    assert in_coords.shape[1] == out_counts.shape[1],  "Input and output arrays must be the same size"
+
+    with nogil:
+        # Define constants which we will use repeatedly.
+        fours = AVX.float_to_float8(4.0)
+        ones = AVX.float_to_float8(1.0)
+        twos = AVX.float_to_float8(2.0)
```

Comment:
This is really good habit! Defining the frequently used variable in the first place and reuse them later... I forgot to do do that and later the codes get really messy...